core/src/main/java/io/grpc/StreamTracer.java

```diff
@@ -62,32 +62,45 @@ public void inboundMessage() {
   }
 
   /**
+   * An outbound message has been completely serialized, revealing its wire size.
+   */
+  public void outboundMessageSerialized(long wireSize) {
```

Comment:
We may be able to implement Option 1 with uncompressed message size, if we can figure out a way to know when the message is fully read out in [SizeEnforcingInputStream](https://github.com/grpc/grpc-java/blob/v1.3.0/core/src/main/java/io/grpc/internal/MessageDeframer.java#L396). @bogdandrutu would that work for you?